### ISSUE TITLE
FileField's Losing Reference

### ISSUE THREAD

**Issue by codeb2cc** [hmarr/mongoengine#458]
FileField's `__get__` method checks instance's `_data` to "Check if a file already exists for this model". If a proxy object does not already exist, after calling `self.fs.put`, since `__get__` returns a brandnew GridFSProxy, the new `grid_id` is then mistakenly overwritten, and the gridfs file reference is forever lost. This means if you don't create a Document with an initial value to the FileField, you can never save gridfs reference to the document anymore(But you DO add new file to MongDB by calling `put`, `replace`, etc).

See https://gist.github.com/1993077

The output is like:

```
==== ==== Test A: Without Initial FileField Value ==== ====
New Document: 4f5760c6be20cc0e17000000
Query the same document: 4f5760c6be20cc0e17000000
Call replace() ...
After replace(), grid_id is None
Save document ...
Query again: 4f5760c6be20cc0e17000000
grid_id is None

==== ==== Test B: With Initial FileField Value ==== ====
New Document: 4f5760c6be20cc0e17000003
Query the same document: 4f5760c6be20cc0e17000003
Call replace() ...
After replace(), grid_id is 4f5760c6be20cc0e17000004
Save document ...
Query again: 4f5760c6be20cc0e17000003
grid_id is 4f5760c6be20cc0e17000004

==== ==== MongoDB GridFS File ==== ====
[u'doc_b', u'doc_e']
```

**Comment by codeb2cc** [comment]
One way to fix this problem:

```
diff --git a/mongoengine/fields.py b/mongoengine/fields.py
index 14fcca0..1389c60 100644
--- a/mongoengine/fields.py
+++ b/mongoengine/fields.py
@@ -853,6 +853,8 @@ class GridFSProxy(object):
             raise GridFSError('This document already has a file. Either delete '
                               'it or call replace to overwrite it')
         self.grid_id = self.fs.put(file_obj, **kwargs)
+        if not self.instance._data.get(self.key):
+            self.instance._data[self.key] = self
         self._mark_as_changed()

     def write(self, string):
```

Another solution (maybe) is to create a default GridFSProxy object when initializing a Document instance, even if the document doesn't actually stored a GridFS file reference.

And I've also noticed that if you've forgot to save the document after put/relace/etc a file, the reference is lost but you do store a new file to MongoDB.

**Comment by rozza** [comment]
Fixed in master - will be in 0.6.8
